rust/src/asn1/parse_rules.rs

```diff
@@ -0,0 +1,270 @@
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use crate::log::*;
+use nom;
+use nom::branch::alt;
+use nom::bytes::complete::tag;
+use nom::character::complete::{digit1, multispace0, multispace1};
+use nom::combinator::{map_res, opt, verify};
+use nom::sequence::{separated_pair, tuple};
+use nom::IResult;
+use std::ffi::CStr;
+use std::os::raw::c_char;
+
+const ASN1_DEFAULT_MAX_FRAMES: usize = 256;
+
+/// Parse the asn1 keyword and return a pointer to a `DetectAsn1Data`
+/// containing the parsed options, returns null on failure
+///
+/// # Safety
+///
+/// pointer must be free'd using `rs_detect_asn1_free`
+#[no_mangle]
+pub unsafe extern "C" fn rs_detect_asn1_parse(input: *const c_char) -> *mut DetectAsn1Data {
+    if input.is_null() {
+        return std::ptr::null_mut();
+    }
+
+    let arg = match CStr::from_ptr(input).to_str() {
+        Ok(arg) => arg,
+        _ => {
+            return std::ptr::null_mut();
+        }
+    };
+
+    match asn1_parse_rule(&arg) {
+        Ok((_rest, data)) => {
+            let mut data = data;
+
+            // Get configuration value
+            if let Some(max_frames) = crate::conf::conf_get("asn1-max-frames") {
+                if let Ok(v) = max_frames.parse::<u16>() {
+                    data.max_frames = v as usize;
+                } else {
+                    SCLogError!("Could not parse asn1-max-frames: {}", max_frames);
+                    return std::ptr::null_mut();
+                };
+            }
+
+            Box::into_raw(Box::new(data))
+        }
+        Err(e) => {
+            SCLogError!("Malformed asn1 argument: {}", e.to_string());
+            std::ptr::null_mut()
+        }
+    }
+}
+
+/// Free a `DetectAsn1Data` object allocated by Rust
+///
+/// # Safety
+///
+/// ptr must be a valid object obtained using `rs_detect_asn1_parse`
+#[no_mangle]
+pub unsafe extern "C" fn rs_detect_asn1_free(ptr: *mut DetectAsn1Data) {
+    if ptr.is_null() {
+        return;
+    }
+    drop(Box::from_raw(ptr));
+}
+
+/// Struct to hold parsed asn1 keyword options
+#[derive(Debug, PartialEq)]
+pub struct DetectAsn1Data {
+    pub bitstring_overflow: bool,
+    pub double_overflow: bool,
+    pub oversize_length: Option<u32>,
+    pub absolute_offset: Option<u16>,
+    pub relative_offset: Option<i32>,
+    pub max_frames: usize,
```

Comment:
Is `usize` like `size_t`, a type with a arch dependent size? I'd prefer an explicit type, like `u32` or `u64`. I *think* we parse this as an `u16`? So then we should use that.

Reply:
No problem! For clarification, in general, should these types be avoided in suricata?

I use `usize` as a counter for recursion here: https://github.com/OISF/suricata/pull/4954/files#diff-d05f7cef7932b801ab5ff5266426ca11R91

Should this be changed?

Reply:
usize is variable in size, 4 bytes on 32 bit, 8 bytes on 32 bit.